sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java

```diff
@@ -0,0 +1,336 @@
+  @Category(RunnableOnService.class)
+  public void testWrite() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NUM_DOCS, InjectionMode.DO_NOT_INJECT_INVALID_DOCS);
+    pipeline
+        .apply(Create.of(data))
+        .apply(ElasticsearchIO.write().withConnectionConfiguration(connectionConfiguration));
+    pipeline.run();
+
+    // force the index to upgrade after inserting for the inserted docs to be searchable immediately
+    node.client().admin().indices().upgrade(new UpgradeRequest()).actionGet();
+
+    SearchResponse response = node.client().prepareSearch().execute().actionGet(5000);
+    assertEquals(NUM_DOCS, response.getHits().getTotalHits());
+
+    QueryBuilder queryBuilder = QueryBuilders.queryStringQuery("Einstein").field("scientist");
+    response = node.client().prepareSearch().setQuery(queryBuilder).execute().actionGet();
+    assertEquals(NUM_DOCS / NUM_SCIENTISTS, response.getHits().getTotalHits());
+  }
+
+  @Rule public ExpectedException exception = ExpectedException.none();
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testWriteWithErrors() throws Exception {
+    TestPipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NUM_DOCS, InjectionMode.INJECT_SOME_INVALID_DOCS);
+    pipeline
+        .apply(Create.of(data))
+        .apply(ElasticsearchIO.write().withConnectionConfiguration(connectionConfiguration));
+    exception.expectCause(isA(IOException.class));
+    exception.expectMessage(
+        new CustomMatcher<String>("RegExp matcher") {
+          @Override
+          public boolean matches(Object o) {
+            String message = (String) o;
+            return message.matches(
+                "(?is).*Error writing to Elasticsearch, some elements could not be inserted"
+                    + ".*document id.*was expecting a colon to separate field name and value"
+                    + ".*document id.*was expecting a colon to separate field name and value.*");
+          }
+        });
+    pipeline.run();
+  }
+
+  @Ignore
+  @Test
+  public void testWriteWithBatchSizes() throws Exception {
+    ElasticsearchIO.Write write =
+        ElasticsearchIO.write()
+            .withConnectionConfiguration(connectionConfiguration)
+            .withMaxBatchSize(BATCH_SIZE);
+    ElasticsearchIO.Write.WriterFn writerFn = spy(new ElasticsearchIO.Write.WriterFn(write));
+    DoFnTester<String, Void> fnTester = DoFnTester.of(writerFn);
+
+    List<String> input = createDocuments(NUM_DOCS, InjectionMode.DO_NOT_INJECT_INVALID_DOCS);
+    //this inserts into Elasticsearch
+    fnTester.processBundle(input);
+    int numWriteBundles = (int) (NUM_DOCS / BATCH_SIZE);
+    // +1 because processBundle calls finishbundle()
+    //TODO do not pass because of TestContext extends OldDoFn.Context whereas Context is in DoFn
+    verify(writerFn, times(numWriteBundles + 1)).finishBundle(Matchers.any(DoFn.Context.class));
```

Comment:
I'd rather recommend to feed it elements one-by-one and verify that every time after you feed BATCH_SIZE elements, the elements get inserted and appear in ES index. That would verify the behavior rather than implementation (black box vs. white box).